src/TypeShape.Examples/MsgPackSerializer/MsgPackSerializer.cs

```diff
@@ -0,0 +1,282 @@
+    }
+
+    /// <summary>
+    /// Deserializes msgpack into a value of a particular type.
+    /// </summary>
+    /// <param name="sequence">The msgpack to deserialize.</param>
+    /// <typeparam name="T">The type of value to deserialize.</typeparam>
+    /// <returns>The deserialized value.</returns>
+    public static T? Deserialize<T>(ReadOnlySequence<byte> sequence)
+        where T : IShapeable<T>
+    {
+        MessagePackReader reader = new(sequence);
+        return SerializerCache<T>.Deserializer(ref reader);
+    }
+
+    private delegate T? Reader<T>(ref MessagePackReader reader);
+
+    private delegate void Writer<T>(ref MessagePackWriter writer, T value);
+
+    private static class SerializerCache<T>
+        where T : IShapeable<T>
+    {
+        private static Writer<T?>? serializer;
+
+        private static Reader<T?>? deserializer;
+
+        internal static Writer<T?> Serializer => serializer ??= SerializeVisitor.GetWriter(T.GetShape());
+
+        internal static Reader<T?> Deserializer => deserializer ??= DeserializeVisitor.GetReader(T.GetShape());
+    }
+
+    private sealed class SerializeVisitor : TypeShapeVisitor
+    {
+        internal static readonly SerializeVisitor Instance = new();
+
+        private static readonly TypeDictionary formatters = new()
+        {
+            { typeof(string), new Writer<string>((ref MessagePackWriter writer, string value) => writer.Write(value)) },
+            { typeof(int), new Writer<int>((ref MessagePackWriter writer, int value) => writer.Write(value)) }
+        };
+
+        internal static Writer<T?> GetWriter<T>(ITypeShape<T> typeShape)
+        {
+            return formatters.GetOrAdd<Writer<T?>>(typeShape, Instance, box => (ref MessagePackWriter writer, T? value) => box.Result(ref writer, value));
+        }
+
+        private SerializeVisitor() { }
+
+        public override object? VisitObject<T>(IObjectTypeShape<T> objectShape, object? state = null)
+        {
+            Dictionary<ReadOnlyMemory<byte>, Writer<T>> properties = new();
+            foreach (IPropertyShape property in objectShape.GetProperties())
+            {
+                if (property.HasGetter/* && property.HasSetter*/)
+                {
+                    byte[] propertyNameBytes = MessagePackSerializer.Serialize(property.Name, MessagePackSerializerOptions.Standard);
+                    properties[propertyNameBytes] = (Writer<T>)property.Accept(this, state)!;
+                }
+            }
+
+            return new Writer<T?>((ref MessagePackWriter writer, T? value) =>
+            {
+                if (value is null)
+                {
+                    writer.WriteNil();
+                    return;
+                }
+
+                writer.WriteMapHeader(properties.Count);
+
+                foreach (var property in properties)
+                {
+                    writer.WriteRaw(property.Key.Span);
+                    property.Value(ref writer, value);
```

Comment:
If `T` is a struct this would force copying of `value` every time a property is being serialized. Consider using a delegate that accepts it by reference.

Reply:
Confirmed that once I change this to `ref`, the `ref` ultimately bleeds out to my top-level `MsgPackSerializer.Serialize` method, which now has to take `ref` instead of the more implicit-friendly `scoped in` parameter.

Reply:
> What do you think? Should/can this be changed to `scoped in` rather than `ref`?

There are circumstances where passing a mutable struct with `in` prompts the codegen to create defensive copies, although I'm not sure if this applies to generic methods. Using ref should result in simpler codegen because it has simpler semantics.

> Confirmed that once I change this to `ref`, the `ref` ultimately bleeds out to my top-level

It's a problem if it leaks to user-facing API. What I usually do to address the issue is author [property-specific delegates/converters](https://github.com/eiriktsarpalis/typeshape-csharp/blob/c8c2c397beb472e716b447bb460dea626811ba18/src/TypeShape.Examples/JsonSerializer/Converters/JsonPropertyConverter.cs#L16-L17) since their signature tends to differ especially for the case of deserialization/binding.

Reply:
I could do that I guess. It means that the Serialize path makes 1 copy of the struct where it could have made 0 though.

Use of generics itself doesn't lead to the defensive copies. It's the C# compiler that does it when invoking a method/property accessor on a struct that is stored in a read-only field but the method isn't labeled `readonly`. 
Consider [this example](https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEBMBGAsAKAKwAIBlAewFsBTAbwOMeMQDs5iAxc8gCleIDCxAG4BDZAFdqASgZN6+JkvQB2MlWoBxanDjUATjzGSZSgNxzGAX2KXbipgBNqyagHNRe4gBUAauKkfbV0DAB5vAXI2UVYDGB9/EwA+PhYfSOjY/WIAYyi4GJYDaQsHRjt0bAggnT19UIF41jgk9RpguuIAXmJUwVzpbtacgDoABX1yAAcDOABPUqsCIgBmYgBnOH0JHPYhBWU15uIJ6dm5oeIAVkWCIA=) where I add some use of generics to make things interesting, but it's basically what I'm proposing for your property getter delegate in that it uses `in` instead of `ref`. Here, because the property getter is manually defined, and lacks a `readonly` modifier, a defensive copy is made before invoking `get_Property`:

```
IL_0000: ldarg.1
IL_0001: ldobj C
IL_0006: stloc.0
IL_0007: ldloca.s 0
IL_0009: call instance int32 C::get_Property()
IL_000e: ret
```

But just add `readonly` to the property, and the same invocation reduces to:

```
IL_0000: ldarg.1
IL_0001: call instance int32 C::get_Property()
IL_0006: ret
```

So we have a choice: always do 1 copy of the struct as part of serialization to avoid any further copies. Or shoot for 0 copies of the struct by using `in` instead of `ref`, but acknowledge you may have *n* copies in reality if you're serializing a struct with custom property getters that don't use `readonly` on them.
Most custom property getter authors probably *don't* bother to add `readonly` to them. So shooting for exactly 1 (by using `ref`) will be faster on average. But if we don't mind shooting for 0 and pointing the blame at others to fix their own code if they complain about excess copies, then using `in` is better.

Reply:
It's probably not the end of the world if only one copy is being made per node (the STJ converters also serialize by value) but if we really wanted to we could make the core delegate/formatter also accept values by ref. In that case we'd be dealing with only one copy for the root-level value in the outer method that resolves the `IShapeable<T>`.

Reply:
In my latest design where I generate `IMessagePackFormatter<T>` (the same currency that MessagePack-CSharp uses), it's a non-issue anyway because we copy structs to serialize them (sadly, but that's the interface). I _do_ use `ref` for internally passing the values around though.